### ISSUE TITLE
Helm find files loses focus

### ISSUE THREAD

**Issue by RidaAyed** [ch11ng/exwm#575]
While browsing with firefox (tridactyl for vim motions), I often need to switch
to some file buffer using `s-<Space> helm-find-files` or `M-x helm-find-files`

In helm, hitting `<tab>` to complete the filepath in the `helm-find-files` buffer,
changes the focus back to firefox and I'm forced to move the mouse, click
around, hit `esc` since the remaining helm buffer becomes unusable.

Do I need to change something related to this <sup><a id="fnr.1" class="footref" href="#fn.1">1</a></sup>line?

#### System Info :computer:

- OS: gnu/linux
- Emacs: 26.2
- Spacemacs: 0.200.13
- Spacemacs branch: master (rev. c7a103a77)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(pdf-tools elfeed mermaid
           (geolocation :variables geolocation-enable-weather-forecast t)
           javascript python mu4e org ranger exwm latex helm auto-completion better-defaults emacs-lisp git markdown
           (shell :variables shell-default-position 'right)
           version-control)
```

##### Footnotes

<sup><a id="fn.1" href="#fnr.1">1</a></sup> <https://github.com/emacs-helm/helm-exwm/blob/master/helm-exwm.el#L191>

**Comment by ch11ng** [comment]
Perhaps pressing `tab` should not select a window displaying a `exwm-mode` buffer.

There is a valid use case where users can activate the minibuffer and navigate to other windows (including those displaying `exwm-mode` buffers) to for example copy something back to the minibuffer.  So for this use case to work, instead of forwarding keys to Emacs whenever the minibuffer window is active we also check if the minibuffer window is currently selected.

**Comment by setzer22** [comment]
@ch11ng  Indeed, pressing <tab> in helm should not select exwm-mode buffers. This currently breaks helm and requires multiple steps to fix. 

Having this working would in turn allow to set helm-follow-mode, which is currently broken inside exwm (because selecting over an exwm-mode buffer selects the exwm buffer and breaks). I think this would be an awesome feature to use helm as a window switcher, because we'd be able to see the window contents, not just the title!

Do you have any directions as to where this should be implemented?

**Comment by ch11ng** [comment]
Strangely I can't seem to reproduce the problem with @RidaAyed 's recipe ATM.  Am I missing something?

> Having this working would allow to set helm-follow-mode.

@setzer22 I'm quite new to helm.  Could you elaborate it?

**Comment by RidaAyed** [comment]
Avoiding `tab` and to continue typing (including `space`) to narrow down to the desired result in helm-mini-buffer works.

**Comment by setzer22** [comment]
@ch11ng Helm can act as a buffer "finder", and has a nice feature called `helm-follow-mode`. In this mode, when the cursor is over a buffer inside helm, the contents of the buffer will be shown in a temporary window. The underlying mechanism for this is the `helm-execute-persistent-action`, which for buffers, shows the contents of the buffer *without changing focus, so helm can still be used*. In `helm-follow-mode`, the `helm-execute-persistent-action` function is executed every time the cursor moves. Also, in spacemacs pressing `<tab>` also executes the same function. 

The nice thing about this feature with exwm-buffers is that you can see the contents of a buffer before you switch to it. A very good use-case for this is when you have multiple windows from the same program and they are not easily distinguishable by its title (e.g. multiple terminal emulator windows).

The problem is that when `helm-execute-persistent-action` is run on an exwm-buffer, the helm window looses focus, and enters in a broken state from where the only way to recover is to kill it. What I expect should happen is helm keeping focus, and the exwm buffer just being displayed.

I first reported this in here because I thought the two issues were closely related. But if you think I'm mistaken, please let me know and I'll submit a new one.

**Comment by ch11ng** [comment]
@setzer22 So here is what I did:

1. `M-x helm-buffers-list RET`
1. Press `C-c C-f` to enable `helm-follow-mode`
1. Use either up or down arrow to preview all buffers

What I found is when an `exwm-mode` buffer is highlighted the minibuffer frame did lose focus.  But up/down arrow keys still worked so it's possible to preview other buffers afterwards.

I've installed the helm and helm-exwm packages.

**Comment by RidaAyed** [comment]
I've made a 16 sec screencast https://imgur.com/a/EqaBfmQ about the issue. Helm focus is lost at 00:00:12 (imgur position 0:05) after hitting TAB. The subsequent keystrokes (up and down arrows) are sent to the other buffer (Firefox) instead.

**Comment by setzer22** [comment]
@ch11ng Doing the exact same thing in my machine won't work. The arrow keys are sent to the `exwm-mode` buffer and helm is frozen until killed. The behaviour I'm experiencing is consistent with the screencast from @RidaAyed.

This makes me think that this may have something to do with either spacemacs or the unofficial exwm layer (https://github.com/timor/spacemacsOS). @RidaAyed, are those two also part of your setup? Maybe that's part of the issue.

**Comment by RidaAyed** [comment]
@setzer22

> are those two also part of your setup?

Indeed. I'm using spacemacs and https://github.com/timor/spacemacsOS

**Comment by ch11ng** [comment]
Could you enable debug mode (`M-x exwm-debug RET`) to prepare a log? Spacemacs makes so many tweaks that it's hard to guess by reviewing configs.

**Comment by setzer22** [comment]
I finally had some time to look into this. I generated the log as requested. This is the sequence of steps that I recorded:
1. In the initial state, I was in the XELB-DEBUG buffer
2. At this point, I pressed `SPC-b-b` which is bound to `helm-mini`
3. In `helm-mini` I highlighted an `exwm-mode` buffer (the Gpick application). 
4. I pressed `TAB` to run `helm-execute-persistent-action`. *NOTE*: Spacemacs rebinds the default for the `TAB` key in helm.
5. The Gpick buffer was shown. Neither window had focus at this point.
6. Using my mouse, I manually focused and closed Gpick. During this whole time, helm was unresponsive.

**Comment by ch11ng** [comment]
@setzer22 The debug log says Gpick got the input focus.  The only key EXWM received was `(65307 . 0)` (a keysym) which is likely an <kbd>escape</kbd> and it was followed by `exwm-reset` (perhaps you bind `exwm-reset` to <kbd>escape</kbd>?).  I'm not sure if the input focus was elsewhere.  Could you execute `export DISPLAY=:0; xwininfo -tree -root; xdotool getwindowfocus` (you'll need to install xdotool and xwininfo) between Step 5 & 6 and see which X window gets the focus? You may either run it from Linux console, or make a new global key binding for it.  Also, when reproducing this issue, try typing something after the input focus is lost.

**Comment by setzer22** [comment]
@ch11ng Just a quick reply (will check with xwininfo later!). But indeed, Gpick or whatever window I try to show gets the input from my keyboard. By "no focus", I actually meant that neither emacs window appeared to be selected.

Anyway, to clarify things, the issue is precisely Gpick getting input focus. The method is supposed to show the buffer but not switch to it. That way you can "peek" a buffer before switching to it. When helm looses focus it becomes impossible to continue browsing buffers.

**Comment by ch11ng** [comment]
I figured it out: that X window got input focus but was in `char-mode` so only global keys worked.  But I'm not sure why it was put into `char-mode`.

**Comment by setzer22** [comment]
All my X windows spawn in `char-mode`. Isn't that the mode where all input is forwarded to the application?

Anyway, I think it may be the code I posted above, which definitely messes with char/line modes. I will test with those changes disabled and see what happens.

**Comment by setzer22** [comment]
After some testing, I've managed to isolate the issue, which is caused by the following lines of code:

```
;; kick all exwm buffers into insert mode per default
(add-hook 'exwm-manage-finish-hook (lambda () (call-interactively #'exwm-input-release-keyboard)))
```
and
```
    ;; ensure that when char mode is left, state is restored to normal
    (advice-add 'exwm-input-grab-keyboard :after (lambda (&optional id)
                                                   (evil-normal-state)))
    ;; ensure that when char mode is entered, input state is activated
    (advice-add 'exwm-input-release-keyboard :after (lambda(&optional id)
                                                      (evil-insert-state)))
```
Disabling all three lines, messes up with spacemacs/exwm in general. However, the related issue seems to be fixed. This led me to investigate further, and I've actually discovered that the issue is only present when the buffer being shown by `helm-execute-persistent-action` is in `char-mode`. If I manually put Gpick in line mode, then I can show its contents without `helm` loosing focus. 

@ch11ng I know this is now probably outside of the scope of exwm. However, any guidance on how to fix this would be much appreciated! My best bet is that we should try to modify `helm-execute-persistent-action` for exwm buffers so that it puts them in char-mode before displaying them. Is there a way to do that in elisp?

**Comment by ch11ng** [comment]
> please, could you try to reproduce the issue with a char-mode buffer to verify my theory?

That's what I did.

> As for a more robust fix for this, do you think there's a way to configure exwm so that running switch-buffer from within helm will not steal the input focus when the input is in char-mode?

In `char-mode` Emacs is not aware of key events targeting other applications.  So to prevent this you have to either use `line-mode`, or make input focus stay on the minibuffer (avoid selecting other Emacs windows or select the minibuffer window again).

**Comment by divansantana** [comment]
I have the same issue. Because my X windows are spawned in `char-mode`. Moreover I had set  `helm-follow-mode-persistent` to true, the default in `nil`. Setting it back to nil, prevents the helm preview and then the issue goes away.

**Comment by codygman** [comment]
I think my issue is related, but doesn't require using helm follow mode or anything like that. Apologies since I don't have anything more specific, but hopefully this helps us get closer to the issue with helm and exwm.

I sometimes hit go to switch buffers quickly after exiting back to line-mode from char-mode and two helm windows open, and I'm in the part where results are usually displayed. Then I can only get around it by doing `C-x o` until I'm in the "correct" helm buffer at the bottom of the screen and hit ctrl-g to exit.

I can then kill all helm related buffers and eventually the correct behavior seems to come back. It's all very odd.

Possibly related to  emacs-helm/helm#194.

It only seems to happen when I've been switching between char-mode and line-mode, then quickly switch to a new buffer.

**Comment by codygman** [comment]
I also noticed there is an issue with company-box where focus is lost while typing normal completions, or the auto-completion window will just disappear. Are the other uses using helm-posframe?

I think company-box also uses a posframe. Does exwm not handle posframe well?
